### PR TITLE
clean up linear interpolation

### PR DESCRIPTION
**Pull request by zingale**
The locate() function was returning the index of the zone that is higher than the input coordinate r.  The logic in interpolate accounted for this, but things were more complex than they need be.  This simplifies locate and the linear interpolation.

It also adds a unit test to help understand what's happening.

## PR checklist

- [ ] test suite needs to be run on this PR
- [ ] this PR will change answers in the test suite to more than roundoff level
- [ ] all newly-added functions have docstrings as per the coding conventions
- [ ] the `CHANGES` file has been updated, if appropriate
- [ ] if appropriate, this change is described in the docs

### CODE REVIEW

**Comment by zingale** [comment]
closes #2663

File: Util/model_parser/model_parser.H [hunk 1]
```diff
@@ -94,18 +98,14 @@ locate(const Real r, const int model_index) {
         loc = ihi;
     }
 
-    return loc;
+    return loc-1;
```
Comment (maxpkatz) [inline Util/model_parser/model_parser.H:101]: What would it mean for locate() to return -1?

Comment (zingale) [reply Util/model_parser/model_parser.H]: should be fixed now.

**Comment by zingale** [comment]
There are now a bunch of unit tests in `Castro/util/model_parser`

These test what happens when you ask for an r outside of the model range and if you get exactly the model data (to roundoff) if you ask for an r which is exactly a model point.

**Comment by zingale** [comment]
okay, looking at this more.  The old interpolation did work fine, in the sense that it had a general enough scheme to check if we need to use i, i-1 or i, i+1 for the bulk of the interpolation.

In fact, the test suite is just roundoff diffs now:

http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/2023-12-17/index.html

File: Util/model_parser/model_parser.H [hunk 2]
```diff
@@ -102,73 +109,35 @@ AMREX_INLINE AMREX_GPU_HOST_DEVICE
...
-            slope = (model::profile(model_index).state(id, var_index) -
-                     model::profile(model_index).state(id-1, var_index)) /
-                (model::profile(model_index).r(id) - model::profile(model_index).r(id-1));
-            interp = slope * (r - model::profile(model_index).r(id)) + model::profile(model_index).state(id, var_index);
-
-        }
-
-    }
+    // safety check to make sure interp lies within the bounding points
+    Real minvar = std::min(model::profile(model_index).state(id+1, var_index),
```
Comment (maxpkatz) [inline Util/model_parser/model_parser.H:126]: Do we want this for r < profile.r(0)?

Comment (zingale) [reply Util/model_parser/model_parser.H:126]: indeed, probably not.  That's a good point.

Comment (zingale) [reply Util/model_parser/model_parser.H]: updated

**Review by maxpkatz — APPROVED** [review]

**Comment by zingale** [comment]
tests are roundoff:
http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/2023-12-26-001/index.html

**Pull request merged**